src/language-js/printer-estree.js

```diff
@@ -1514,6 +1514,19 @@ function printPathNoParens(path, options, print, args) {
         const needsForcedTrailingComma =
           canHaveTrailingComma && lastElem === null;
 
+        const shouldBreak =
+          n.elements.length > 1 &&
+          n.elements.filter(
+            element =>
+              element &&
+              ((element.type === "ArrayExpression" &&
+                element.elements &&
+                element.elements.length > 1) ||
+                (element.type === "ObjectExpression" &&
+                  element.properties &&
+                  element.properties.length > 1))
+          ).length > 1;
```

Comment:
Let’s see if I understand this. If there are at least two “special” items in an array, it will always break. A “special” item is an array/object with at least 2 things in it. Is that correct? So `[1, 2, [3, 4], {a, b}]` would break?

@thorn0 is this the logic you were envisioning?

Reply:
Not exactly. My proposal was to break only if all the elements are special as the point of this change is to improve the readability of homogeneous arrays of arrays and arrays of objects. I'm not even sure about mixed arrays like `[[1, 2], { a, b }]`. Mixed arrays look readable enough without breaking to me.

Reply:
@thorn0 Do you think the following case should be broken?
```js
[[0], [0, 1], [0, 1]];
[{ a }, { a, b }, { a, b }];
```

Reply:
Good question. I'd say let's not break for now and wait for feedback.

Reply:
@sosukesuzuki BTW this is going to solve https://github.com/prettier/prettier/issues/5780.